api/src/services/taxonomy-service.ts

```diff
@@ -1,43 +1,94 @@
+   */
+  async _elasticSearch(
+    searchRequest: SearchRequest
+  ): Promise<SearchResponse<ITaxonomySource, Record<string, AggregationsAggregate>> | undefined> {
     try {
       const client = new Client({ node: process.env.ELASTICSEARCH_URL });
       return client.search({
-        index: 'taxonomy',
+        index: process.env.ELASTICSEARCH_TAXONOMY_INDEX,
         ...searchRequest
       });
     } catch (error) {
       defaultLog.debug({ label: 'elasticSearch', message: 'error', error });
     }
   }
 
-  private sanitizeSpeciesData = (data: SearchHit<any>[]) => {
-    return data.map((item) => {
-      const label = [
-        item._source.code,
-        [
-          [item._source.tty_kingdom, item._source.tty_name].filter(Boolean).join(' '),
-          [item._source.unit_name1, item._source.unit_name2, item._source.unit_name3].filter(Boolean).join(' '),
-          item._source.english_name
+  /**
+   * Determines if the source object from a taxonomy search query is valid (namely, that it has not expired).
+   * @param source The source object from the search query
+   * @returns {boolean} `true` if the the `end_date` property is not defined or if the date hasn't passed, `false` otherwise.
+   */
+  _isValidTaxonomySource = (source: ITaxonomySource | undefined): boolean => {
+    if (source?.end_date) {
+      return new Date() < new Date(source.end_date);
+    }
+
+    return true;
+  };
+
+  /**
+   * Sanitizes species data retrieved from Elasticsearch.
+   * @param {SearchHit<ITaxonomySource>[]} data The data response fromEelasticsearch
+   * @returns {{ id: string, label: string }} An ID and label pair for each taxonomic code
+   */
+  _sanitizeSpeciesData = (data: SearchHit<ITaxonomySource>[]): { id: string; label: string }[] => {
+    return data
+      .filter((item: SearchHit<ITaxonomySource>) => this._isValidTaxonomySource(item._source))
+      .map((item: SearchHit<ITaxonomySource>) => {
+        const { _id: id, _source } = item;
+
+        const label = [
+          _source?.code,
+          [
+            [_source?.tty_kingdom, _source?.tty_name].filter(Boolean).join(' '),
+            [_source?.unit_name1, _source?.unit_name2, _source?.unit_name3].filter(Boolean).join(' '),
+            _source?.english_name
+          ]
+            .filter(Boolean)
+            .join(', ')
         ]
           .filter(Boolean)
-          .join(', ')
-      ]
-        .filter(Boolean)
-        .join(': ');
+          .join(': ');
 
-      return { id: item._id, label: label };
-    });
+        return { id, label };
+      });
   };
 
+  /**
+   * Searches the taxonomy Elasticsearch index by taxonomic code IDs
+   * @param ids The array of taxonomic code IDs
+   * @returns The response from Elasticsearch
+   */
   async getTaxonomyFromIds(ids: number[]) {
-    const response = await this.elasticSearch({
+    const response = await this._elasticSearch({
       query: {
         terms: {
           _id: ids
```

Comment:
Is there any way to handle the filtering through some sort of elastic search concept, so we don't have to filter the results after fetching them?

Reply:
There is actually - currently working on implementing it